### ISSUE TITLE
The simulation result of running baseline is 0（drag、lift、Rec Area all is 0）

### ISSUE THREAD

**Issue by Brightmoonc** [jerabaul29/Cylinder2DFlowControlDRL#30]
Hi~
Following the tutorial you gave me, I successfully installed the singularity container, and when I started Launching one episode without training-without controlling the flow, the results showed：

Singularity> python3 single_runner.py 
iUFL can be obtained from https://github.com/MiroK/ufl-interpreter
resume env
--- call init ---
mesh_file = mesh/turek_2d.h5
-0.2 0.21
total_number_of_probes = 151
--- done init ---
define network specs
define agent
environment.states = {'shape': (151,), 'type': 'float'}
environment.actions = {'min_value': -0.01, 'max_value': 0.01, 'shape': (2,), 'type': 'float'}
network_spec = [{'size': 512, 'type': 'dense'}, {'size': 512, 'type': 'dense'}]
2024-09-18 02:11:19.425992: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2024-09-18 02:11:19.485192: E tensorflow/stream_executor/cuda/cuda_driver.cc:406] failed call to cuInit: CUDA_ERROR_UNKNOWN
2024-09-18 02:11:19.485875: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:145] kernel driver does not appear to be running on this host (ubuntu): /proc/driver/nvidia/version does not exist
Trained Network not found...
start simulation
2024-09-18 02:11:20.147293: W tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc:1441] Failed to build SimpleGraphView.
2024-09-18 02:11:20.148654: E tensorflow/core/grappler/optimizers/dependency_optimizer.cc:586] Iteration = 0, topological sort failed with message: Non-existent input ^ConstantFoldingCtrl/ppo/cond/add/Switch_0 for node ConstantFolding/ppo/cond/strided_slice_1/stack_const_axis
2024-09-18 02:11:20.149597: E tensorflow/core/grappler/optimizers/dependency_optimizer.cc:586] Iteration = 1, topological sort failed with message: Non-existent input ^ConstantFoldingCtrl/ppo/cond/add/Switch_0 for node ConstantFolding/ppo/cond/strided_slice_1/stack_const_axis
/usr/lib/python3/dist-packages/matplotlib/quiver.py:645: RuntimeWarning: divide by zero encountered in double_scalars
  length = a * (widthu_per_lenu / (self.scale * self.width))
/usr/lib/python3/dist-packages/matplotlib/quiver.py:645: RuntimeWarning: invalid value encountered in multiply
  length = a * (widthu_per_lenu / (self.scale * self.width))
/usr/lib/python3/dist-packages/matplotlib/quiver.py:701: RuntimeWarning: invalid value encountered in less
  short = np.repeat(length < minsh, 8, axis=1)
/usr/lib/python3/dist-packages/matplotlib/quiver.py:711: RuntimeWarning: invalid value encountered in less
  tooshort = length < self.minlength
/usr/lib/python3/dist-packages/matplotlib/backend_bases.py:2437: MatplotlibDeprecationWarning: Using default event loop until function specific to this GUI is implemented
  warnings.warn(str, mplDeprecation)
Simu_lift | Ep N:    1, step:    0, Rec Area: 0.0000, drag: 0.0000, lift: 0.0000
Simu_lift | Ep N:    1, step:  100, Rec Area: 0.0000, drag: 0.0000, lift: 0.0000
Simu_lift | Ep N:    1, step:  200, Rec Area: 0.0000, drag: 0.0000, lift: 0.0000
...
Single Run finished. AvgDrag : 0.0, AvgRecircArea : 0.0

I don't know why the simulation results are all 0, I didn't make any changes to the other files. （I am running on a vmware workstation-ubuntu system）
I am a beginner in this field, and I think you are very strong in this field after reading your article in JFM, so I would like to start by replicating your results in this field. I would appreciate it very much if you could tell me the reason for the above problem

**Comment by jerabaul29** [comment]
Hi,

I think the error is due to:

>  Trained Network not found...

i.e. you are trying to run a single runner with NN controller (which requires a NN for the controller), but you do not actually have a trained NN. Not sure then if either the controller is random, or if it just outputs NaNs - but the result is the same, this is not the expected use, and it does not work.

I have not used this code in a long time, so I do not have it fresh in memory, but I think there is a specific single runner you should use; from the readme:

> Go in Cylinder2DFlowControlWithRL/baseline_flow. Just launch single_runner.py to look at the baseline flow. During the computation, the graphical interface shows the state of the flow, and the data is saved in baseline_flow/results for the vtu files, and in baseline_flow/saved_models for the csv files.

this is not the same single runner as in the Cylinder2DFlowControlWithRL/ANN_controlled_flow_learning folder.

Remember that all these commands must be executed from inside the container, and within the code that is available there - not in "just" a clone of the present repository.

This has been used successfully by many students in several different groups in the past, so I feel confident this is a "end user issue", probably not using the code in the right folder / not using the code from inside the container. Closing for now, if you have more questions we can discuss further.

**Comment by Brightmoonc** [comment]
Thank you for your response!
But what I just used is the specific single runner in the baseline_flow folder, and I am sure it's running inside the container.

> Singularity> ls
ANN_controlled_flow_learning   baseline_flow	 geometry_2d.template_geo
ANN_controlled_flow_singlerun  empty_simulation  jet_bcs.py
Env2DCylinder.py	       flow_solver.py	 msh_convert.py
__pycache__		       generate_msh.py	 probes.py
Singularity> cd baseline_flow
Singularity> python3 single_runner.py


Are the problems shown below the cause?
> 2024-09-18 02:11:19.425992: I tensorflow/core/platform/cpu_feature_guard.cc:140] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2024-09-18 02:11:19.485192: E tensorflow/stream_executor/cuda/cuda_driver.cc:406] failed call to cuInit: CUDA_ERROR_UNKNOWN
2024-09-18 02:11:19.485875: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:145] kernel driver does not appear to be running on this host (ubuntu): /proc/driver/nvidia/version does not exist
Trained Network not found...
start simulation
2024-09-18 02:11:20.147293: W tensorflow/core/grappler/optimizers/arithmetic_optimizer.cc:1441] Failed to build SimpleGraphView.
2024-09-18 02:11:20.148654: E tensorflow/core/grappler/optimizers/dependency_optimizer.cc:586] Iteration = 0, topological sort failed with message: Non-existent input ^ConstantFoldingCtrl/ppo/cond/add/Switch_0 for node ConstantFolding/ppo/cond/strided_slice_1/stack_const_axis
2024-09-18 02:11:20.149597: E tensorflow/core/grappler/optimizers/dependency_optimizer.cc:586] Iteration = 1, topological sort failed with message: Non-existent input ^ConstantFoldingCtrl/ppo/cond/add/Switch_0 for node ConstantFolding/ppo/cond/strided_slice_1/stack_const_axis

**Comment by jerabaul29** [comment]
Ok, strange, I would think that this kind of errors should not happen when using a singularity container that abstracts most of the low level stuff.

- what kind of machine are you on? can you try on another machine / a laptop or similar?
- can you try to re-run everything from scratch? i.e. run first `make_mesh.py`, then run `perform_learning.py` a bit (just to re-generate the baseline locally if needed), then run `single_runner.py`? you can try this both from the `baseline_flow` folder, and from the `empty_simulation` folder, to see if re-doing everything from scratch helps

Not really sure what is happening - I and many other people I know have used this with good success quite a lot in the past. But this container is 5 years old, so some stuff may be incompatible somehow with newer versions of something.

Otherwise, you can also look into:

https://github.com/venturi123/DRLinFluids

this is doing similar tasks, is quite a bit more recent, and I think the group that was the main driver behind it is still actively supporting it :) .
